### PR TITLE
feat: relax extensions file structure validation

### PR DESCRIPTION
**Pull request by smira**
* allow empty directories (I see no harm in having them)
* allow symlinks 

See also https://github.com/talos-systems/extensions/pull/20

Signed-off-by: Andrey Smirnov <andrey.smirnov@talos-systems.com>

---
This change is [Reviewable](https://reviewable.io/reviews/talos-systems/talos/5136)

### CODE REVIEW

File: internal/pkg/extensions/validate.go [hunk 1]
```diff
@@ -72,7 +72,16 @@ func (ext *Extension) validateContents() error {
...
 		if d.Type().Type() == os.ModeSymlink {
-			return fmt.Errorf("symlinks are not allowed: %q", itemPath)
+			var dest string
+
+			dest, err = os.Readlink(path)
+			if err != nil {
+				return err
+			}
+
+			if strings.IndexByte(dest, '/') != -1 || dest == "." || dest == ".." {
```
Comment (frezbo) [inline internal/pkg/extensions/validate.go:82]: a symlink like this could potentially skip this test `ln -s ./../notes.txt foo`?

**Review by frezbo — CHANGES_REQUESTED** [review]

File: internal/pkg/extensions/validate.go (hunk 1 above)
Comment (smira) [reply internal/pkg/extensions/validate.go:82]: nope, as we don't allow slashes for now, so symlink could only be a filename like `a.so` -> `a.so`

File: internal/pkg/extensions/validate.go [hunk 2]
```diff
@@ -88,24 +83,10 @@ func (ext *Extension) validateContents() error {
 		}
 
 		// no special files
-		if !d.IsDir() && !d.Type().IsRegular() {
+		if !d.IsDir() && !d.Type().IsRegular() && d.Type().Type() != os.ModeSymlink {
```
Comment (frezbo) [inline internal/pkg/extensions/validate.go:86]: `d.Type().Type()` intentional?

**Review by frezbo — CHANGES_REQUESTED** [review]

File: internal/pkg/extensions/validate.go (hunk 2 above)
Comment (smira) [reply internal/pkg/extensions/validate.go:86]: and it's covered by tests!

**Review by frezbo — APPROVED** [review]

**Pull request merged**